dbms/src/IO/Compression/CompressionCodecRunLength.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <IO/Compression/CompressionCodecRunLength.h>
+#include <IO/Compression/CompressionInfo.h>
+#include <IO/Compression/CompressionSettings.h>
+#include <IO/Compression/EncodingUtil.h>
+#include <IO/Compression/ICompressionCodec.h>
+#include <common/unaligned.h>
+#include <lz4.h>
+
+#include <magic_enum.hpp>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_COMPRESS;
+extern const int CANNOT_DECOMPRESS;
+} // namespace ErrorCodes
+
+CompressionCodecRunLength::CompressionCodecRunLength(CompressionDataType data_type_)
+    : data_type(data_type_)
+{}
+
+UInt8 CompressionCodecRunLength::getMethodByte() const
+{
+    return static_cast<uint8_t>(CompressionMethodByte::RunLength);
+}
+
+UInt32 CompressionCodecRunLength::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    return 1 + LZ4_COMPRESSBOUND(uncompressed_size);
+}
+
+template <typename T>
+UInt32 CompressionCodecRunLength::compressDataForInteger(const char * source, UInt32 source_size, char * dest) const
+{
+    constexpr auto bytes_size = sizeof(T);
+    if unlikely (source_size % bytes_size != 0)
+        throw Exception(ErrorCodes::CANNOT_DECOMPRESS, "source size {} is not aligned to {}", source_size, bytes_size);
+    const char * source_end = source + source_size;
+    DB::Compression::RunLengthPairs<T> rle_vec;
+    rle_vec.reserve(source_size / sizeof(T));
+    for (const auto * src = source; src < source_end; src += sizeof(T))
+    {
+        T value = unalignedLoad<T>(src);
+        if (rle_vec.empty() || rle_vec.back().first != value
+            || rle_vec.back().second == std::numeric_limits<UInt8>::max())
+            rle_vec.emplace_back(value, 1);
+        else
+            ++rle_vec.back().second;
```

Comment:
Why do we need `rle_vec.back().second == std::numeric_limits<UInt8>::max()` in the if branch?

Reply:
`255 + 1 = 0`

Reply:
Add comments about this situation